### PR TITLE
Adds HACKING.md for newbies

### PR DESCRIPTION
**Pull request by UtkarshBhatthere**

### CODE REVIEW

File: HACKING.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+```bash
+sudo snap install snapcraft --classic
+sudo snap install lxd
+```
+
+> **_NOTE:_**
+For a detailed how-to-use snapcraft tool guide, check-out [Snap-Tutorials](https://snapcraft.io/docs/snap-tutorials)
+
+## 📖 References
+The microceph codebase resides inside the microceph sub-directory of the repo. This subdir is neatly organised into parts which make up the whole of microceph. Below is a brief description of what can be found in each of those parts for rederence.
```
Comment (sabaini) [inline HACKING.md:30]: s/rederence/reference/

File: HACKING.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+### Sub Directories
+1. **[api](/microceph/api)**
+
+    Contains files related to the internal REST APIs which the microceph client uses to communicate to the microceph daemon. It also contains a types subdir which has necessary data structures for the APIs.
+
+2. **[ceph](/microceph/ceph)**
+
+    Contains files which directly related to ceph orchestration. This includes code for ceph cluster configuration, service orchestration etc.
```
Comment (sabaini) [inline HACKING.md:39]: s/which directly/which are directly/

File: HACKING.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    Contains common code
+
+6. **[database](/microceph/database)**
+
+    Contains DQlite schema and migration definitions along with generated mappers for DB interfacing.
+
+7. **[mocks](/microceph/mocks)**
+
+    Contains mock definitions for Unit Testing
```
Comment (sabaini) [inline HACKING.md:59]: s/Unit Testing/unit testing/

File: HACKING.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+```bash
+# Locally built snaps do no auto-connect the available plugs on install, they can be connected manually using;
+sudo snap connect microceph:block-devices
+sudo snap connect microceph:hardware-observe
+sudo snap connect microceph:dm-crypt
+sudo snap restart microceph.daemon
+```
+
+## 👍 Unit-Testing
+The microceph [makefile](/microceph/Makefile) has targets for running Unit Tests and Lint checks. They can be run as follows.
```
Comment (sabaini) [inline HACKING.md:88]: s/Unit Testing/unit testing/
s/Lint/lint/

**Review by sabaini — CHANGES_REQUESTED** [review]
Hey, thanks for this, looks very nice.

One general naming suggestion on MicroCeph vs. microceph which comes up in several places. To lessen the confusion around CamelCase vs lowercase in naming IMHO we almost always should refer to MicroCeph (as in the MicroCeph product), and only refer to the lowercase microceph when we are specifically talking about the built microceph snap
file, or `microceph` subdirs or similar.

Some typo nits inline.

File: HACKING.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# Microceph Hacking Guide
+Hold your horses, no-one's hacking anything. The aim of this document is to enable new developers/users get familiarised with the build tools and codebase of microceph so that they can build and contribute to the [Microceph Codebase](https://github.com/canonical/microceph).
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply HACKING.md:2]: Aaaah, the good hackers :D.. Ack

File: HACKING.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+1. [⚡️ Introduction to snaps](#⚡️-introduction-to-snaps)
+2. [🧰 Tools](#🧰-tools)
+3. [📖 References](#📖-references)
+4. [🏭 Build Guide](#🏭-build-guide)
+5. [👍 Unit-Testing](#👍-unit-testing)
+
+## ⚡️ Introduction to snaps
+> A [snap](https://snapcraft.io/about) is a bundle of an app and its dependencies that works without modification across Linux distributions.
+
+Apart from being a self dependant artifact, being a snap enables Microceph to be isolated from host and provide a clean and consistent building, installing and cleanup experience.
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply HACKING.md:14]: Ack

File: HACKING.md (hunk 1 above)
Comment (UtkarshBhatthere) [reply HACKING.md:30]: Nice Catch :D

File: HACKING.md (hunk 2 above)
Comment (UtkarshBhatthere) [reply HACKING.md:39]: Ack

File: HACKING.md (hunk 4 above)
Comment (UtkarshBhatthere) [reply HACKING.md:88]: ack

File: HACKING.md (hunk 3 above)
Comment (UtkarshBhatthere) [reply HACKING.md:59]: ack

**Review by sabaini — APPROVED** [review]

**Pull request merged**